Add support for double tap and drag for text selection

Renzo-Olivares opened the pull request:
This change adds support for double tap + drag, to select word by word when double tapping and then dragging the mouse.

This change also adds a new gesture recognizers that will be specifically useful for text selection:
* `TapAndDragGestureRecognizer`

This recognizer makes use of a `_TapStatusTrackerMixin` to keep track of the tap count.

`TextSelectionGestureDetector` no longer uses a `TapGestureRecognizer` or `PanGestureRecognizer`. `TapUp`,`TapDown`, `TapCancel`, `onStart`, `onUpdate`, `onEnd`, and `onCancel` are now handled by the new recognizer.

Currently our text selection system does not support double tap + drag (selects word by word) and triple tap + drag (selects line by line). This PR adds support for this behavior by bundling the handling of tap up and tap down with a drag gesture recognizer and long press gesture recognizer. This helps us avoid having to pass state around to accomplish the same behavior. For example, this behavior could also be accomplished by having the TapGestureRecognizer.onTapDown callback, save the tap count, then this tap count can be used by the DragGestureRecognizer.onUpdate callback to determine if the selection should be updated character by character, word by word, or line by line. Some disadvantages of passing state around between gesture recognizers are that you can't accomplish this behavior in a classes initializer (Trying to build a default mapping of gesture recognizers for text selection https://github.com/Renzo-Olivares/flutter/blob/3dbffa3922c656e5c2f97fb6ff8d42b4573d152c/packages/flutter/lib/src/widgets/default_selection_gestures.dart#L24).

Will be handled in separate PR: 
Selection area #104552

Fixes #99071, #64550, #23973

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

Commit:
Replace PanGestureRecognizer in TextSelection with TapAndDragGestureRecognizer

Commit:
add tracking of _DragState to new tap_and_drag recognizer and remove some legacy double tap code from text_selection.dart and add logs"

Commit:
add dragTapCount, a tap count that is persistent for an entire drag and is set to null on drag end vs the regular tap count which is reset on a timer

Commit:
basic double tap to drag functionality and add a local dragTapCount in text_selection.dart to use with the timer callback

Commit:
Add offsetFromOrigin and localOffsetFromOrigin to DragUpdateDetails similar to LongPressMoveUpdateDetails, eliminates the need to hold the state of lastDragStartDetails

Commit:
make a generic baselongpressgesturerecognizer

Commit:
Revert "make a generic baselongpressgesturerecognizer"

This reverts commit aad8f7433bd01e4cd016d527af832c3b1f15fac5.

Commit:
rename tap_and_drag to selection_recognizers

Commit:
add mixin for consecutivetap

Commit:
tap and long press gesture recognizer

Commit:
Revert "Revert "make a generic baselongpressgesturerecognizer""

This reverts commit 181350c36718f644eada3e45c1b7b5939f90a340.

Commit:
Revert "Revert "Revert "make a generic baselongpressgesturerecognizer"""

This reverts commit 4d69775967858dfd66dd9429e1713da598908a85.

Commit:
Add support for secondary button clicks on drag gesture recognizer and separate drag end and tap up callback

Commit:
get test running

Commit:
rename tapCount to consecutiveTapCount

Commit:
dispose timer properly

Commit:
add some comments to tests

Commit:
Add comments

chunhtai commented on packages/flutter/lib/src/gestures/selection_recognizers.dart:
```diff
@@ -0,0 +1,514 @@
+
+  void _checkLongPressStart() {
+    print('from recognizer check long press start');
+    if (_isDoubleTap) {
```
does this work for double tap and drag?

Renzo-Olivares replied:
Do you mean if we can do this case for double tap and drag?, or if this case breaks double tap to drag?

chunhtai reviewed (commented):
I only take a quick glance at the code. I like the direction this is going, thanks for refactoring it.

Renzo-Olivares commented on packages/flutter/lib/src/gestures/drag_details.dart:
```diff
@@ -191,6 +195,16 @@ class DragUpdateDetails {
   /// Defaults to [globalPosition] if not specified in the constructor.
   final Offset localPosition;
 
+  /// A delta offset from the point where the drag initially contacted
```
Would this be break since `offsetFromOrigin` has a default value, and `localOffsetFromOrigin` is nullable? One of the reasons i'm in favor of adding this, is that `LongPressMoveUpdateDetails` provides a similar API https://github.com/flutter/flutter/blob/9cb269c194b02d6b4688698460e4947f9f270ea0/packages/flutter/lib/src/gestures/long_press.dart#L166. Also that would avoid us having to save the `DragStartDetails` in our current text selection implementation, which is something I can't do with the text gestures project since the gesture recognizer mapping is defined in the classes initializer.

Renzo-Olivares commented on packages/flutter/lib/src/gestures/selection_recognizers.dart:
```diff
@@ -0,0 +1,514 @@
+  }
+
+  void consecutiveTapTimeout() {
+    print('consecutive tap timeout');
```
Will remove all, just have for debugging right now.

Commit:
Make ConsecutiveTapMixin private and move logic to increment tap count into mixin

Commit:
stop tracking pointer when gesture is rejected and detect drags on touch devices

Commit:
onCancel for TapAndDrag

Commit:
have the TapAndDragGestureRecognizer handle tap downs and tap ups on touch and mouse devices

Commit:
add drag to move cursor for android and iOS, and pointer device kind to DragUpdateDetails

Commit:
get tests running

Commit:
refactor TapAndDragGestureRecognizer moving some logic into _check methods

Commit:
Handle cancel properly on TapAndDragGestureRecognizer, having both onTapCancel and onDragCancel, also fix tests

Commit:
Fix test mouse drag selects and cannot drag cursor, save _initialPosition based on dragStartBehavior (either on tapDown or dragStart)

Commit:
determine if drag has a sufficient global distance to accept and fix some cancel behavior, making _checkCancel clearer

Commit:
give up pointer on drag end

Commit:
properly stop tracking pointer, fixes test for right click on Apple and non-apple platforms

Commit:
clean up some comments from last commit

Commit:
remove drag on touch for now

Commit:
fix Can select text by dragging with a mouse due to dragStart only being fired on the first PointerMoveEvent, the previous pan gesture recognizer would fire both dragStart and dragUpdate

Commit:
Revert "fix Can select text by dragging with a mouse due to dragStart only being fired on the first PointerMoveEvent, the previous pan gesture recognizer would fire both dragStart and dragUpdate"

This reverts commit 124dc79bc3389672c76d7c014ce04edab297abc6.

Commit:
correctly use _initialPosition for checkStart and call _checkUpdate after _checkStart if localDelta is not zero

Commit:
updates

Commit:
fix double tap chains

Commit:
Add docs

Commit:
Address analyzer

Commit:
more analyzer, only issues left are with print statements

Commit:
add deadlineTimer to fix conflict with ForcePressGestureRecognizer

Commit:
Revert "add deadlineTimer to fix conflict with ForcePressGestureRecognizer"

This reverts commit 3b29ddfff4cde4845edd481ecefb789fea2a0781.

Commit:
remove unecessary changes to tests

Commit:
secondaryButton should not drag

Commit:
Revert "Revert "add deadlineTimer to fix conflict with ForcePressGestureRecognizer""

This reverts commit 0a008f029f5796acd48c17c1897c0b700d5ef3a7.

Commit:
updates

Commit:
Revert "updates"

This reverts commit 4803b8443a2b67f0b8d29e9a01f712dfcb0f588c.

Commit:
Revert "Revert "Revert "add deadlineTimer to fix conflict with ForcePressGestureRecognizer"""

This reverts commit 79251a7af88d5dbb1460a960afc77e65dea18bff.

Commit:
fix shift + tap + drag tests, this was happening because a double tap + drag was being registered and not a single tap, added a duration to pumpAndSettle to fix this

Commit:
remove TapAndLongPressGestureRecognizer

Commit:
fix cupertino text field tests related to shift + tap + drag

Commit:
deadline timer try 2

Commit:
more logs

Commit:
Should reset taps when tap cancel is called, and should wait until gesture is accepted to initiate a drag

Commit:
should clear _down and _up when gesture is rejected

Commit:
remove erroneous log

Commit:
fix selectable text double tap chains test

Commit:
dont restart timer until tap up

Commit:
reset consecutiveTapCount on drag end

Commit:
fix selectableText test

Commit:
fix material text field tests

Commit:
reject TapAndDragGestureRecognizer when it is neither a tap nor a drag

Commit:
remove prints

Commit:
clean up

Commit:
shift aware

Commit:
clean up

Commit:
fix cupertino test

Commit:
fix text field focus tests

Commit:
Add 100ms delay to cupertino test, to prevent a double tap

Commit:
clean up test comments

Commit:
add comment to test

Commit:
uncomment test

Commit:
remove longpress changes

Commit:
Fix drag on mobile

Commit:
remove debug

Commit:
Fix drag to move cursor on iOS

Commit:
left over from drag fix

Commit:
add tests for drag on touch devices

Commit:
add test for double tap + drag mouse devices

Commit:
add tests

Commit:
Fix bug where initialPosition was used before it was set

justinmc commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -1634,6 +1621,31 @@ class TextSelectionGestureDetectorBuilder {
   }
 
+  /// Returns true if position was on selection.
+  bool _lastTapWasOnSelection(Offset position, TextSelection? targetSelection) {
```
Nit: Maybe change the name of the method since `position` doesn't necessarily have to be the last tap?

justinmc commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -1634,6 +1621,31 @@ class TextSelectionGestureDetectorBuilder {
+        && targetSelection.end >= textPosition.offset;
+  }
+
+  bool _tapWasOnSelection(Offset position) {
```
Same for this about the name of the method.

justinmc commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -1758,21 +1758,17 @@ class TextSelectionGestureDetectorBuilder {
       case TargetPlatform.fuchsia:
       case TargetPlatform.iOS:
         // On mobile platforms the selection is set on tap up.
-        if (_isShiftTapping) {
```
It's really cool to see all of this shifttapping stuff go away and be replaced by a normal gesture detector :+1:

justinmc commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -2127,63 +2117,137 @@ class TextSelectionGestureDetectorBuilder {
       final Offset startOffset = renderEditable.maxLines == 1
           ? Offset(renderEditable.offset.pixels - _dragStartViewportOffset, 0.0)
           : Offset(0.0, renderEditable.offset.pixels - _dragStartViewportOffset);
+      final Offset dragStartGlobalPosition = details.globalPosition - details.offsetFromOrigin;
```
Is this something that you could always do to get the start position?  It seems nice that we don't need DragStartDetails any more.

Renzo-Olivares replied:
Yes! because I added `offsetFromOrigin` to `DragUpdateDetails`. This is similar to `LongPressMoveUpdateDetails` which also has `offsetFromOrigin`. A good use case where that is currently used is in TextField https://github.com/flutter/flutter/blob/e3c29ec00c9c825c891d75054c63fcc46454dca1/packages/flutter/lib/src/material/text_field.dart#L76 .

justinmc commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -2127,63 +2117,137 @@ class TextSelectionGestureDetectorBuilder {
           : Offset(0.0, renderEditable.offset.pixels - _dragStartViewportOffset);
+      final Offset dragStartGlobalPosition = details.globalPosition - details.offsetFromOrigin;
+
+      // Select word by word.
```
I didn't realize this but even with a mouse this is right, it selects by word for me here on Linux.

justinmc commented on packages/flutter/test/material/text_field_test.dart:
```diff
@@ -2135,6 +2135,98 @@ void main() {
     expect(controller.selection.extentOffset, testValue.indexOf('g'));
   });
 
+
```
Too many newlines here.

justinmc commented on packages/flutter/test/cupertino/text_field_test.dart:
```diff
@@ -3446,7 +3446,7 @@ void main() {
       await tester.tapAt(endpoints[0].point + const Offset(1.0, 1.0));
       await tester.pump();
-      await tester.pump(const Duration(milliseconds: 200)); // skip past the frame where the opacity is zero
+      await tester.pump(const Duration(milliseconds: 300)); // skip past the frame where the opacity is zero
```
Is there a constant somewhere that could be used in all of the cases where you use 300ms to make it more clear where this comes from?

Renzo-Olivares replied:
Yeah I can do that 300ms is the kDoubleTapTimeout from gestures/constants.dart

justinmc reviewed (commented):
Overall I'm super excited about having a built-in gesture recognizer for tap and drag.  Thanks for cleaning all of this up.  Most comments are smaller things.

Commit:
Address some review comments and fix issue where if double tap was held too long then long press gesture recognizer would take over

Commit:
remove _isDoubleTap flag since it is no longer needed due to previous commit

Commit:
Add docs for onTapCancel and onDragCancel

Commit:
analyzer fixes

Commit:
Do not test selection handles on macOS, since macOS does not support touch

Commit:
Add assert for dragStartBehavior

Commit:
add double tap + drag tests to cupertino

Commit:
use kDoubleTapTimeout instead of const Duration(milliseconds: 300) for readability

Commit:
analyzer issues

Renzo-Olivares commented on packages/flutter/lib/src/gestures/selection_recognizers.dart:
```diff
@@ -0,0 +1,662 @@
+    this.dragStartBehavior = DragStartBehavior.start,
+    super.kind,
+    super.supportedDevices,
+  });
```
Added the assert!

Renzo-Olivares commented on packages/flutter/lib/src/gestures/selection_recognizers.dart:
```diff
@@ -0,0 +1,662 @@
+      // If last tap offset is false then we have restarted our consecutive tap count,
+      // so the consecutiveTapTimer should be null.ß
+      assert(consecutiveTapTimer == null);
+      consecutiveTapCount += 1;
```
Corrected this

Commit:
update docs

Commit:
update more docs

Commit:
address comments

Commit:
more doc updates

Commit:
fix docs

Commit:
unused import

Commit:
fix docs

Commit:
Add more tests

Commit:
Add more tests and reject a tap up if we have exceeded the tap tolerance

Commit:
updates

Renzo-Olivares commented:
@justinmc @chunhtai @LongCatIsLooong This is ready for review. 

There's an analyzer issue due to a dependency loop that happens when I bring in `HardwareKeyboard.instance` from `services.dart` into my `selection_recognizers.dart`. I can move `selection_recognizers` into the `widgets` folder, since my plan was for it to live in the same file as the selection gestures widget I'm working on, that would fix this dependency loop. However, I'd like to get everyones take on handing the `Shift Tapping` in the recognizer itself, or if that logic should be delegated somewhere else. 

My main motivation for having it in the recognizer is so that I can then define a map of GestureRecognizer subclasses to GestureRecognizerFactory, in the initialization of a class https://github.com/Renzo-Olivares/flutter/blob/3dbffa3922c656e5c2f97fb6ff8d42b4573d152c/packages/flutter/lib/src/widgets/default_selection_gestures.dart#L43.

chunhtai commented on packages/flutter/lib/src/gestures/drag_details.dart:
```diff
@@ -185,21 +192,36 @@ class DragUpdateDetails {
+  /// A delta offset from the point where the drag initially contacted
+  /// the screen to the point where the pointer is currently located (the
+  /// present [globalPosition]) when this callback is triggered.
+  final Offset offsetFromOrigin;
```
Which origin should this be if this is a double tap and drag? may be good to document this

Renzo-Olivares replied:
It would be the global position of the most recent tap down, i.e. the second tap down of a double tap and drag.

chunhtai commented on packages/flutter/lib/src/gestures/drag_details.dart:
```diff
@@ -185,21 +192,36 @@ class DragUpdateDetails {
   final Offset localPosition;
 
+  /// A delta offset from the point where the drag initially contacted
+  /// the screen to the point where the pointer is currently located (the
```
```suggestion
  /// the screen to the point where the pointer is currently located in global coordinates (the
```

chunhtai commented on packages/flutter/lib/src/gestures/drag_details.dart:
```diff
@@ -185,21 +192,36 @@ class DragUpdateDetails {
+  /// present [globalPosition]) when this callback is triggered.
+  final Offset offsetFromOrigin;
+
+  /// A local delta offset from the point where the drag initially contacted
```
maybe
```
A delta offset from the point where the drag initially contacted the screen to the point where the pointer is currently located in local coordinates relative to <TextField?>
```

Renzo-Olivares replied:
I don't think we should add the TextField specification since these Details object is used outside of TextField.

chunhtai replied:
That is just place holder. When it say local coordinates, it only makes sense if it mention what it is relative to.

chunhtai commented on packages/flutter/lib/src/gestures/drag_details.dart:
```diff
@@ -139,14 +141,19 @@ class DragUpdateDetails {
     required this.globalPosition,
+    this.kind,
     Offset? localPosition,
+    this.offsetFromOrigin = Offset.zero,
```
Is this necessary? In the past we let the caller to calculate this if they need.

tangential to this pr, I just realized there are states in  the drag gesture reconginzer, will the internal state messed up if it is sharing between multiple textfields?

Renzo-Olivares replied:
I've done this to avoid having to save `lastStartDetails` and passing that to the drag update method to calculate the origin of the drag.

`LongPressMoveUpdateDetails` has a similar API, and we also use it in text selection. https://github.com/flutter/flutter/pull/109573#discussion_r975775224

I think it's a nice to have, to avoid saving the`DragStartDetails`.

chunhtai commented on packages/flutter/lib/src/gestures/selection_recognizers.dart:
```diff
@@ -0,0 +1,714 @@
+/// The consecutive tap count when the drag was initiated is given by [TapStatus.consecutiveTapCount].
+///
+/// Used by [TapAndDragGestureRecognizer.onUpdate].
+typedef GestureDragUpdateWithTapStatusCallback = void Function(DragUpdateDetails details, TapStatus status);
```
Since the TapStatus is specific to selection gesture, consider create a special DragUpdateDetails? I am not sure what is long term plan for these detail class @dkwingsmt may have better suggestion.

chunhtai commented on packages/flutter/lib/src/gestures/selection_recognizers.dart:
```diff
@@ -0,0 +1,714 @@
+  ///
+  /// Can be null to indicate that the gesture can drift for any distance.
+  /// Defaults to [kTouchSlop].
+  final double? preAcceptSlopTolerance;
```
not entirely sure why this was exposed in other gesture recongizer, but you can use doc template to avoid duplicated the doc string from

chunhtai commented on packages/flutter/lib/src/gestures/selection_recognizers.dart:
```diff
@@ -0,0 +1,714 @@
+  ///
+  /// Can be null to indicate that the gesture can drift for any distance.
+  /// Defaults to [kTouchSlop].
+  final double? postAcceptSlopTolerance;
```
same

chunhtai commented on packages/flutter/lib/src/gestures/selection_recognizers.dart:
```diff
@@ -0,0 +1,714 @@
+  /// {@macro flutter.gestures.tap.TapGestureRecognizer.onSecondaryTapUp}
+  GestureTapUpCallback? onSecondaryTapUp;
+
+  /// {@macro flutter.gestures.monodrag.DragGestureRecognizer.onStart}
```
may need to document the tap status but, here and other places

chunhtai commented on packages/flutter/lib/src/gestures/selection_recognizers.dart:
```diff
@@ -0,0 +1,714 @@
+      kind: getKindForPointer(event.pointer),
+    );
+
+    incrementConsecutiveTapCountOnDown(details.globalPosition);
```
I imagine the _ConsecutiveTapMixin to be more self contain. can the _ConsecutiveTapMixin to be mix on OneSequenceGestureRecognizer so that it can override addAllowGesture, accept gesture and the rejectGesture to update its own state? TapAndDragGestureRecognizer can just query the consecutiveTapCount when the drag starts

Renzo-Olivares replied:
I think we have to continue to query the consecutiveTapCount in the onTapDown. The primary reason being there can be a double click hold and then drag. If we query during drag start the consecutiveTapCount may have been reset by the `Timer` timeout callback.

chunhtai replied:
you can override handleEvent to get record most recent down event?

chunhtai commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -2444,60 +2475,32 @@ class _TextSelectionGestureDetectorState extends State<TextSelectionGestureDetec
         widget.onSingleLongTapEnd != null) {
       gestures[LongPressGestureRecognizer] = GestureRecognizerFactoryWithHandlers<LongPressGestureRecognizer>(
-        () => LongPressGestureRecognizer(debugOwner: this, kind: PointerDeviceKind.touch),
+        () => LongPressGestureRecognizer(debugOwner: this, supportedDevices: <PointerDeviceKind>{ PointerDeviceKind.touch }),
```
Can this also be replaced by TapAndDragGestureRecognizer?

Renzo-Olivares replied:
At the moment I did not see a reason to expand the new recognizer's capabilities to support `LongPress`. I haven't identified any selection behaviors that rely on consecutive taps for long press.

chunhtai replied:
how do we handle longpress and drag?

chunhtai reviewed (commented):
This looks great, I left some comments on the selection_gesture APIs

Renzo-Olivares commented on packages/flutter/lib/src/cupertino/text_field.dart:
```diff
@@ -101,7 +101,7 @@ class _CupertinoTextFieldSelectionGestureDetectorBuilder extends TextSelectionGe
 
   @override
-  void onSingleTapUp(TapUpDetails details) {
+  void onSingleTapUp(TapUpDetails details, TapStatus status) {
```
That wouldn't include the `isShiftTapping` field (status of shift being pressed).

Commit:
Address comments

justinmc commented on packages/flutter/lib/src/gestures/selection_recognizers.dart:
```diff
@@ -0,0 +1,701 @@
+  final int consecutiveTapCount;
+
+  /// Whether the shift key was pressed when this tap happened.
+  final bool isShiftPressed;
```
Nit: Maybe `shiftWasDown` or something like that?  I don't know if I like that any better...  But it's slightly confusing that this means that the shift was down at the time of the tap, not necessarily when it is accessed.

Renzo-Olivares replied:
I changed this to `keysPressedOnDown` in case there are other keyboard + gesture combinations that need to be handled in the future. I have so far identified one other gesture besides shift, which is ctrl. ctrl + click acts like a right click on macOS, selecting the word at the clicked position and popping up the context menu.

justinmc replied:
Ah good call!

justinmc reviewed (commented):
The one thing that I want to double check is the existence of TapStatus separate from the gesture Details.  It just seems strangely different from the other gesture callbacks when I see both parameters.  Could it be combined into the relevant Details classes, or is this approach actually better so we don't have to make changes to so many Details classes?

Renzo-Olivares commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -2127,63 +2104,137 @@ class TextSelectionGestureDetectorBuilder {
+          to: details.globalPosition,
+          cause: SelectionChangedCause.drag,
+        );
+      }
```
Yes triple tap and drag, will be handled in another PR.

Commit:
fix test naming

Renzo-Olivares commented on packages/flutter/test/cupertino/text_field_test.dart:
```diff
@@ -1835,6 +1835,103 @@ void main() {
   );
 
+  testWidgets(
+    'double tap + drag selects word by word on mouse devices',
```
Changed to
`Can double tap + drag to select word by word` and
`Can double click + drag with a mouse to select word by word`

Commit:
update documentation

Commit:
move selection_recognizers to selection_gestures

Commit:
fix analyzer

Renzo-Olivares commented:
> The one thing that I want to double check is the existence of TapStatus separate from the gesture Details. It just seems strangely different from the other gesture callbacks when I see both parameters. Could it be combined into the relevant Details classes, or is this approach actually better so we don't have to make changes to so many Details classes?

I see what you mean. It feels like `TapStatus` should just be a part of the gesture `Details` object. I think some new `Details` object might be the way forward here, since `TapGestureRecognizer` and `DragGestureRecognizer` won't make use of the consecutive tap count or shift status.

Commit:
fix analyzer

Commit:
keysPressedOnDown instead of isShiftPressed

Commit:
update docs

Commit:
update docs

Commit:
Add drag update throttle to TapAndDragGestureRecognizer

Commit:
update comments

Commit:
missed from merge

Commit:
Replace _ConsecutiveTapMixin with _TapStatusTrackerMixin

Commit:
updates

Commit:
correctly cancel tap when when past tap tolerance with new implementation

Commit:
Should call tap and drag cancel if we are giving up a pointer without succesfully tracking a PointerUpEvent

Commit:
comments

Commit:
move pastTapTolerance to tap tracker

Commit:
move pastTapTolerance to tap tracker

Commit:
clean up check for nulls and remove use of consecutiveTapCountWhileDragging

Commit:
move call to super.acceptGesture to top

Commit:
remove print

Commit:
clean up

Commit:
Fix tests where both PanGestureRecognizer and TapAndDragGestureRecognizer lost

Commit:
clean up

Commit:
_GestureState -> _DragState

Commit:
more docs clean up

Commit:
more clean up

Commit:
Add onSecondaryTapCancel

Commit:
Add docs

Commit:
more docs

Renzo-Olivares commented on packages/flutter/lib/src/gestures/selection_recognizers.dart:
```diff
@@ -0,0 +1,662 @@
+
+/// Recognizes taps and movements.
+///
+/// Takes on the responsibilities of [TapGestureRecognizer] and [DragGestureRecognizer] in one [GestureRecognizer].
```
Yes you can

Commit:
Fix broken isPointerAllowed when attempting a right click drag - the _initialButtons is never reset

Commit:
revert debug flag

Renzo-Olivares commented:
This is ready for another review. Most of the changes since the last review happened in selection_gestures.dart. They mostly include _ConsecutiveTapTrackerMixin -> _TapStatusTrackerMixin. And the _TapStatusTrackerMixin is now more self-sufficient in setting it’s own state.

Commit:
make primaryPointer private

Commit:
Add support for upper count limit in TapAndDragGestureRecognizer, the tap counter should not be allowed to grow infinitely unless that is desired

Commit:
fix analyzer

Commit:
Use new TapDrag details objects and callbacks

Commit:
clean up docs

Commit:
clean up and add test for upperLimit

justinmc commented on packages/flutter/test/cupertino/text_field_test.dart:
```diff
@@ -3625,6 +3726,12 @@ void main() {
+    // This way it can declare itself the winner before the [TapAndDragGestureRecognizer] that
+    // is on the selection overlay.
+    await tester.pump();
+    await gesture.moveTo(handlePos - toNextLine);
```
I ran your branch locally to double check that my handle dragging stuff still works correctly and it does 👍 I can't tell a difference between now and before this PR.  So these changes to the handle dragging tests are fine with me.

justinmc commented on packages/flutter/test/cupertino/text_field_test.dart:
```diff
@@ -3966,6 +4073,97 @@ void main() {
+    // Tap on text field to gain focus, and set selection to '|d'. On iOS
+    // the selection is set to the word edge closest to the tap position.
+    // We await for 300ms after the up event, so our next down event does not
+    // register as a double tap.
```
Nit: Should you use kDoubleTapTimeout instead?

justinmc replied:
Ah I see that I mentioned this in an old comment below.  Still holds I think.

Renzo-Olivares replied:
I may have missed a few, just went back and changed the rest.

justinmc commented on packages/flutter/test/widgets/selection_gestures_test.dart:
```diff
@@ -0,0 +1,370 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
```
Thanks for creating this test file.

justinmc reviewed (commented):
Overall looks good to me, mostly small comments.  I checked out your branch and tried it out on iOS and Mac and it works great.  Vertical handle dragging still works perfectly as before.

Renzo-Olivares commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -2058,25 +2052,22 @@ class TextSelectionGestureDetectorBuilder {
   ///    this callback.
   @protected
-  void onSingleTapUp(TapUpDetails details) {
+  void onSingleTapUp(TapDragUpDetails details) {
```
I think this definitely has potential to be a breaking change

Renzo-Olivares replied:
I just did a search internally and it doesn't look like anyone is extending `TextSelectionGestureDetector` just using it in `TextField`.

Renzo-Olivares commented on packages/flutter/lib/src/widgets/selection_gestures.dart:
```diff
@@ -0,0 +1,1251 @@
+          invokeCallback('onSecondaryTapDown', () => onSecondaryTapDown!(details));
+        }
+        break;
+      default:
```
Yes because we should start tracking `_initialButtons` in `addAllowedPointer`. If `_initialButtons` is null at this point we cannot determine which callback to fire.

Renzo-Olivares replied:
I added a few asserts here in case.

Renzo-Olivares commented on packages/flutter/lib/src/widgets/selection_gestures.dart:
```diff
@@ -0,0 +1,1251 @@
+///  * [TapDragUpDetails], the details for [GestureTapDragUpCallback].
+///  * [TapDragStartDetails], the details for [GestureTapDragStartCallback].
+///  * [TapDragEndDetails], the details for [GestureTapDragEndCallback].
+class TapDragUpdateDetails {
```
I thought about it, I actually tried extending them initially but wasn't super happy how it looked. I'm not sure in what direction to go with that but am open to suggestions. Currently none of the details objects extend each other so I followed that pattern. I think it might be a nice-to-have to refactor them to use a common base class, but also wasn't sure if I wanted to add more to an already large PR.

justinmc replied:
That sounds good to me.  I'll keep it in mind as we move forward with this in case any idea strikes me, but it definitely makes sense as-is.

Renzo-Olivares commented on packages/flutter/lib/src/widgets/selection_gestures.dart:
```diff
@@ -0,0 +1,1251 @@
+  /// the number in the series this tap is.
+  final int consecutiveTapCount;
+
+  /// The keys that were pressed when the most recent [PointerDownEvent] occurred.
```
Exactly. I think it's fine that the `keysPressedOnDown` represents the keys down of the most recent tap. I would think a gesture would seem a bit unintuitive if it was something like on the 2nd tap of a quadruple tap if shift was down then do some action on the 4th tap.

justinmc replied:
Yeah that sounds fine to me :+1:

Renzo-Olivares commented on packages/flutter/lib/src/widgets/selection_gestures.dart:
```diff
@@ -0,0 +1,1251 @@
+  final PointerDeviceKind kind;
+
+  /// If this tap is in a series of taps, then this value represents
+  /// the number in the series this tap is.
```
That is correct, if you continue to tap again after a drag then the tap series is reset. I'll put together some more docs.

Renzo-Olivares commented on packages/flutter/lib/src/widgets/selection_gestures.dart:
```diff
@@ -0,0 +1,1251 @@
+
+  @override
+  void acceptGesture(int pointer) {
+    _wonArena = true;
```
I'm not sure I follow. Do you place the `_wonArena` in an else/ after `if` and `if`? 

I think it's fine to set `_wonArena` where we do since `_consecutiveTapTimerStop/Start` don't make use of it. It's mostly if our recognizer declares victory before `handleEvent`, in this case we will not have tracked a `PointerUpEvent` yet so we can't start the timer on `acceptGesture`, so we wait until the `handleEvent` following the acceptance to track our `PointerUpEvent` and start the timer.

justinmc replied:
Sorry that sentence is really confusing now that I re-read it haha.  This is what I meant, though that makes sense and it's totally fine keeping it where it is:

```dart
    if (_up != null && _down != null) {
      _consecutiveTapTimerStop();
      _consecutiveTapTimerStart();
      _wonArena = false;
    } else {
      _wonArena = true;
    }
```

or

```dart
    if (_up != null && _down != null) {
      _consecutiveTapTimerStop();
      _consecutiveTapTimerStart();
      _wonArena = false;
      return;
    }
    _wonArena = true;
```

Renzo-Olivares replied:
I think I'll leave it as is. 

The first example will only set `_wonArena` to true only if `_up` or `_down` is null, I think this is not accurate because the arena is won at this point regardless if we have an `_up` or `_down`. 

The second example is similar to what I have now, though I think I like setting `_wonArena` to true first.

Renzo-Olivares commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -1819,6 +1807,23 @@ class TextSelectionGestureDetectorBuilder {
+      return false;
+    }
+
+    final TextPosition textPosition = renderEditable.getPositionForPoint(position);
```
I think all should be fine, but I can take a look and see if i can make any improvements. In reality we don't need to call this method on every drag update, since it is always using the same `dragStartGlobalPosition` and `_dragStartSelection` that was tracked at the beginning of the drag in `onStart`. We should be able to call this once, and then it should be true for the remaining drag update events.

Renzo-Olivares replied:
Fixed this by using a new flag `dragBeganOnPreviousSelection`, it is set on drag update only if it is null by calling `_positionOnSelection` which calls `getPositionForPoint` in question.  Here it is called once for the entirety of the drag. It is reset to null on drag end.

Renzo-Olivares commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -1942,11 +1935,16 @@ class TextSelectionGestureDetectorBuilder {
+    // in renderEditable. The gesture callbacks can use the details objects directly
+    // in callbacks variants that provide them [TapAndDragGestureRecognizer.onSecondaryTap]
+    // vs [TapAndDragGestureRecognizer.onSecondaryTapUp] instead of having to track state in
+    // renderEditable. When this migration is complete we should remove this hack.
```
I will make one.

Renzo-Olivares replied:
https://github.com/flutter/flutter/issues/115130

Commit:
Add docs for TapAndDragGestureRecognizer and remove some ambiguity of onStart onUpdate and onEnd parameters

Commit:
Address review comments

Commit:
analyzer fixes

Renzo-Olivares commented on packages/flutter/test/cupertino/text_field_test.dart:
```diff
@@ -2124,6 +2124,105 @@ void main() {
+      expect(controller.selection.baseOffset, testValue.indexOf('d'));
+      expect(controller.selection.extentOffset, testValue.indexOf('i') + 1);
+    },
+  );
```
I don't think we have any tests for `TextSelectionGestureDetector` inside of `EditableText` since it is only available at the `TextField` level. The only gesture related tests I see on `EditableText` are ones that make use of the internal gesture recognizers `RenderEditable` has. I think we should keep these tests where they are for now, but definitely explore how we can make the situation better. I always find it a bit cumbersome to have to write tests x2 sometimes x3 if we include `SelectableText`.

Since we want our new Text selection gestures to follow platform behaviors versus follow design language behaviors (material vs cupertino), then I agree the best path forwards are to move all the selection tests to one place since they shouldn't differ between cupertino and material.

justinmc replied:
Ah that's right, sounds good, but agreed that we should find a way to consolidate these in the future.

Commit:
Call cancel before rejecting the gesture so we can still access _initialButtons

chunhtai commented on packages/flutter/lib/src/widgets/selection_gestures.dart:
```diff
@@ -0,0 +1,1320 @@
+/// Used by [TapAndDragGestureRecognizer.onTapDown].
+typedef GestureTapDragDownCallback  = void Function(TapDragDownDetails details);
+
+/// Details for [GestureTapDragDownCallback], such as the tap count within
```
```suggestion
/// Details for [GestureTapDragDownCallback], such as how many continuous taps before the drag starts
```

chunhtai commented on packages/flutter/lib/src/widgets/selection_gestures.dart:
```diff
@@ -0,0 +1,1320 @@
+  // all tap related state is reset and a new tap series is tracked.
+  //
+  // If this value is null, [consecutiveTapCount] can grow infinitely large.
+  int? get upperLimit;
```
maybe call `maxConsecutiveTap` ?

chunhtai commented on packages/flutter/lib/src/widgets/selection_gestures.dart:
```diff
@@ -0,0 +1,1320 @@
+    if (_up != null && _down != null) {
+      _consecutiveTapTimerStop();
+      _consecutiveTapTimerStart();
+      _wonArena = false;
```
What does the _wonArena parameter really mean? How can a gesturereconginzer un-win itself without caliing rejectGesture?

Renzo-Olivares replied:
It means the gesture recognizer has won the arena for the given pointer sometime in the past. This is used because `acceptGesture` might be called before `handleEvent` and in that case we will not have tracked a `PointerUpEvent` yet so we cannot call tap up until we track the `PointerUpEvent` in `handleEvent`. So I use `_wonArena` to communicate this to `handleEvent`.

Renzo-Olivares replied:
Yes because we want to start the timer on tap up and stop it on tap down. Better explained here: https://github.com/flutter/flutter/pull/109573#discussion_r1041485695

chunhtai commented on packages/flutter/lib/src/widgets/selection_gestures.dart:
```diff
@@ -0,0 +1,1320 @@
+
+  @override
+  void handleEvent(PointerEvent event) {
+    if (event is PointerMoveEvent) {
```
Why does this need to check for move? we should only need to worry about the slop distance when the pointer is up?

Renzo-Olivares replied:
There can be a `PointerMoveEvent` that occurs between a `PointerDownEvent` and `PointerUpEvent`. See https://github.com/flutter/flutter/blob/52b3dc25f6471c27b2144594abb11c741cb88f57/packages/flutter/lib/src/gestures/recognizer.dart#L611

chunhtai commented on packages/flutter/lib/src/widgets/selection_gestures.dart:
```diff
@@ -0,0 +1,1320 @@
+      if (_dragState == _DragState.possible) {
+        // If we arrive at a [PointerUpEvent], and the recognizer has not won the arena, and the tap drift
+        // has exceeded its tolerance, then we should reject this recognizer.
+        if (pastTapTolerance) {
```
Shouldn't this be handled by the mixin?

chunhtai commented on packages/flutter/lib/src/widgets/selection_gestures.dart:
```diff
@@ -0,0 +1,1320 @@
+          _giveUpPointer(event.pointer);
+          return;
+        }
+        // The drag has not been accepted before a [PointerUpEvent], therefore the recognizer
```
I can't wrap my head around the logic.
Is it using `pastTapTolerance` to infer the drag has not been accepted?

Renzo-Olivares replied:
Say we have a sequence of events `PointerDownEvent` -> `PointerMoveEvent` -> `PointerUpEvent`. There are 3 cases.

If we arrive at that `PointerUpEvent` and the `PointerMoveEvent` did not move a sufficient global distance to be accepted as a drag, but moved a sufficient global distance to move past the tap tolerance.
- TapCancel
- DragCancel

If we arrive at that `PointerUpEvent` and the `PointerMoveEvent` moved a sufficient global distance to be accepted as a drag, therefore also moving past the tap tolerance because dragTolerance > tapTolerance.
- TapCancel

If we arrive at that `PointerUpEvent` and the `PointerMoveEvent` did not move a sufficient global distance to be accepted as a drag and it did not move past the tap tolerance.
- DragCancel

chunhtai replied:
> the pointer did not move a sufficient global distance to be accepted as a drag.

In this case, why does it need to send DragCancel? Does the Drag already start at this point?

Renzo-Olivares replied:
It looks like in case #1 I am not matching the target behavior. Where individual `TapGestureRecognizer` and `DragGestureRecognizer` will detect the case as a drag, my single `TapAndDragGestureRecognizer` rejects both tap and drag.

Renzo-Olivares replied:
I am now mostly matching the behavior in scenario `#1`. 

With two recognizers `TapGestureRecognizer` and `PanGestureRecognizer`:
`onDown-PanGestureRecognizer` -> `onStart-PanGestureRecognizer` -> `onUpdate-PanGestureRecognizer` -> `onEnd-PanGestureRecognizer`

With one `TapAndDragGestureRecognizer`:
`onTapDown-TapAndDragGestureRecognize` -> `onTapCancel-TapAndDragGestureRecognize` -> `onDragStart-TapAndDragGestureRecognize` -> `onDragUpdate-TapAndDragGestureRecognize` -> `onDragEnd-TapAndDragGestureRecognize`.

The difference is that the new recognizer calls `onTapDown` and `onTapCancel` followed by `onDragStart` and `onDragEnd`. This differing behavior is expected for the new recognizer since it will call `onTapDown` when `acceptGesture` is called, and that needs to be followed up with an `onTapCancel` when a drag begins. 

When using two recognizers (`TapGestureRecognizer` and `PanGestureRecognizer`), `onTapDown` is not called because `TapGestureRecognizer` only fires it on `acceptGesture` and `TapGestureRecognizer` self declares loss when moving past tap tolerance, therefore there is also no need for it to call `onTapCancel` since tap down was never fired. In this situation `PanGestureRecognizer` wins and the gesture is pan.

Our case differs because `TapGestureRecognizer` declares loss, which gives `acceptGesture` in `TapAndDragGestureRecognizer` the chance to call `onTapDown`, but because our pointer moved past the tap tolerance this gesture cannot be a tap, so we call tap cancel and the gesture defaults to a drag.

chunhtai replied:
It feels weird that we merge two gesture recognizers to eliminate the conflicting callback and cancelling each other, but now we still simulating the canceling behavior in one merged gesture recognizer. In my opinion, once the canceling is called, this gesture recognizer should not fire any other callback in this gesture sequence. I think we should treat tap and drag as one kind of gesture so in this case 
`onTapDown-TapAndDragGestureRecognize -> onTapCancel-TapAndDragGestureRecognize -> onDragStart-TapAndDragGestureRecognize -> onDragUpdate-TapAndDragGestureRecognize -> onDragEnd-TapAndDragGestureRecognize.`

it doesn't make sense to call `onTapCancel-TapAndDragGestureRecognize`. The `onTapDown` can be a down for a single tap or a down for a drag. The cusumer (i.e editabletext) should have this in mind when using this recognizer. If they have to differentiate it, they need to implement their own logic or keep some kind of state.

Does this make sense?

Renzo-Olivares replied:
This makes a lot of sense! I think my initial train of thought regarding the decision on cancel behavior was to keep some tests happy, but it makes a lot more sense to change the intended behavior and have the recognizer only have one cancel.

Renzo-Olivares replied:
Updated the cancel behavior to only cancel when a tap down has been fired, i.e. when a PointerDownEvent was received. Also merged `onTapCancel` and `onDragCancel` into `onCancel`. I agree that the `onCancel` should be the callback fired in a sequence of callbacks of the recognizer.

chunhtai reviewed (commented):
It looks like the mixin still not separating the functionality well enough that the `TapAndDragGestureRecognizer` and `_TapStatusTrackerMixin` needs to use fragmented information to guess the status of each other's state.

Is it possible to come up with a mechanism in _TapStatusTrackerMixin to start or stop tracking the tap? something like `_TapStatusTrackerMixin.startTrackingTap` or `_TapStatusTrackerMixin.stopTrackingTap`. The `TapAndDragGestureRecognizer` will just look for when drag is accepted.

Before drag is accepted, everything will be handled by `_TapStatusTrackerMixin`. Once drag is accepted, `TapAndDragGestureRecognizer` call `_TapStatusTrackerMixin.stopTrackingTap` and start handling the event.

for `_TapStatusTrackerMixin` it will just handle event as long as `_TapStatusTrackerMixin.stopTrackingTap` is not called.

Will this make thing cleaner? If this is not possible, we may consider just merge them together since they are too inner-connected

Commit:
Recognizer should reject any pointer differing from the original

Renzo-Olivares commented on packages/flutter/lib/src/widgets/selection_gestures.dart:
```diff
@@ -0,0 +1,1320 @@
+  String toString() => '${objectRuntimeType(this, 'TapDragEndDetails')}($velocity)';
+}
+
+// A mixin for [OneSequenceGestureRecognizer] that tracks the number of taps
```
Will trigger doc generation, or no since this is still a private class?

chunhtai replied:
I think it will but i am not sure.

Renzo-Olivares replied:
In that case I think maybe we should keep it `//` since `_TapStatusTracker` is private.

chunhtai replied:
sounds good to me

Renzo-Olivares commented on packages/flutter/lib/src/widgets/selection_gestures.dart:
```diff
@@ -0,0 +1,1320 @@
+    required this.consecutiveTapCount,
+    required this.keysPressedOnDown,
+  }) : assert(globalPosition != null),
+       localPosition = localPosition ?? globalPosition;
```
Will do.

Renzo-Olivares replied:
Done.

Commit:
Revert "Recognizer should reject any pointer differing from the original"

This reverts commit afd9807480bd11e119bdd2b7d520631511973bab.

Commit:
Address reviewer comments

Commit:
Correct cancel behavior

Commit:
Fix consecutive tap + drag because _dragStart state was not being set when consecutive tap is greater than one

Commit:
Add more tests

Commit:
Add documentation on behavior with TapGestureRecognizer and DragGestureRecognizer

Commit:
more docs

Commit:
more docs

Commit:
remove comments

Commit:
updates

Commit:
fix multiple pointer behavior

Commit:
only handle the primary pointer

Commit:
Clean up dangerous assumptions in gesture details objects

Commit:
forgot from rebase

Commit:
update docs

Commit:
updates

Commit:
Clean up some redundant code

Commit:
remove whitespace

Commit:
fix tests as a result of #115849

Commit:
update test docs

Commit:
Fix same test from last commit for material variants

Commit:
More clean up of redundant code and update docs

Commit:
Clean up didStopTrackingLastPointer and untie TapAndDragGestureRecognizer cancel behavior from TapStatusTrackerMixin.currentUp state

Commit:
untie pastTapTolerance

Commit:
updates

Commit:
Add slopTolerance

Commit:
update docs

Renzo-Olivares commented on packages/flutter/lib/src/widgets/selection_gestures.dart:
```diff
@@ -0,0 +1,1383 @@
+  ///
+  ///  * [kSecondaryButton], the button this callback responds to.
+  ///  * [onSecondaryTapUp], which has the same timing but with details.
+  GestureTapCallback? onSecondaryTap;
```
You are right, I think it just makes sense to use `TapGestureRecognizer` for those responsibilities then. It also cleans up the code a bit. Thanks!

Commit:
Have secondary tap handled by TapGestureRecognizer

Commit:
update docs

Renzo-Olivares commented on packages/flutter/lib/src/widgets/selection_gestures.dart:
```diff
@@ -0,0 +1,1383 @@
+      }
+    } else if (event is PointerUpEvent) {
+      _up = event;
+      if (_wonArena && _up != null && _down != null) {
```
This goes hand in hand with line 982 https://github.com/Renzo-Olivares/flutter/blob/tap_drag_gestures/packages/flutter/lib/src/widgets/selection_gestures.dart#L982 in the`TapAndDragGestureRecognizer` where we try to call tap up in the case when we have previously won the arena before a tap up was received.

The consecutive tap timer should start on tap up and is stopped on the following tap down. 

To accomplish this, because the mixin is the one handling the tap count logic, I have tried to line the mixin and recognizer up to run in a sequence. The mixin runs first, doing its own tap tracker related logic to the `PointerEvent` received (doing its own check tap up and check tap down basically), and then the recognizer runs its own logic on the same `PointerEvent`. This order can be assured because `super` is called first in all inherited methods, so the mixin runs first.

Renzo-Olivares replied:
To add on `_wonArena` is needed for the same reason the recognizer uses it (to call tap up if we previously won the arena).

Commit:
fix analyzer and address comments

Commit:
Add more docs

Commit:
Update cancel behavior tol not call on tap cancel when a drag has been accepted

Commit:
Change cancel behavior to only cancel if the tap down callback has been sent and merge tapcancel and dragcancel

Commit:
update docs;

Renzo-Olivares commented on packages/flutter/lib/src/widgets/selection_gestures.dart:
```diff
@@ -0,0 +1,1383 @@
+  void acceptGesture(int pointer) {
+    _wonArena = true;
+    if (_up != null && _down != null) {
+      _consecutiveTapTimerStop();
```
This section of the code in `_TapStatusTrackerMixin.acceptGesture` goes in sequence with `TapAndDragGestureRecognizer.acceptGesture`. The mixin runs first then the recognizer. The code in the mixin checks if there is a tap up, if there is then it starts the timer.

Event order:

PointerUpEvent received
_TapStatusTrackerMixin.acceptGesture - checks tap up by checking if we tracked a previous PointerUpEvent and starts timer if we have
TapAndDragGestureRecognizer.acceptGesture - calls its own tap up.

This ties into my explanation https://github.com/flutter/flutter/pull/109573#discussion_r1041485695

Renzo-Olivares commented on packages/flutter/lib/src/widgets/selection_gestures.dart:
```diff
@@ -0,0 +1,1383 @@
+  /// Can be null to indicate that the gesture can drift for any distance.
+  /// Defaults to 18 logical pixels.
+  @override
+  double? slopTolerance;
```
Do you mean something like

`TapAndDragGestureRecognizer({ Duration? deadline, ..}) : _deadline = deadline;` ?

chunhtai replied:
just use final unless you plan to change it internally in this class

Commit:
Rename selection_gestures to tap_and_drag_gestures

Renzo-Olivares commented on packages/flutter/lib/src/widgets/selection_gestures.dart:
```diff
@@ -0,0 +1,1383 @@
+    this.dragStartBehavior = DragStartBehavior.start,
+    this.dragUpdateThrottleFrequency,
+    this.maxConsecutiveTap,
+    this.slopTolerance = kTouchSlop,
```
I'm on board with this still need to change it.

justinmc commented on packages/flutter/lib/src/widgets/tap_and_drag_gestures.dart:
```diff
@@ -0,0 +1,1296 @@
+
+  /// If this tap is in a series of taps, then this value represents
+  /// the number in the series this tap is.
+  final int consecutiveTapCount;
```
Nit: You could use a template for repeated docs like this if you want.

justinmc commented on packages/flutter/lib/src/widgets/tap_and_drag_gestures.dart:
```diff
@@ -0,0 +1,1296 @@
+
+  @override
+  void rejectGesture(int pointer) {
+    _tapTrackerReset();
```
No need to call super here?

Renzo-Olivares replied:
`OneSequenceGestureRecognizer` leaves its implementations for `acceptGesture` and `rejectGesture` empty so I don't think we need to call them in this case.

justinmc commented on packages/flutter/lib/src/widgets/tap_and_drag_gestures.dart:
```diff
@@ -0,0 +1,1296 @@
+/// When competing against [TapGestureRecognizer], if the pointer does not move past the tap
+/// tolerance, then the recognizer that entered the arena first will win. In this case the
+/// gesture detected is a tap. If the pointer does travel past the tap tolerance then this
+/// recognizer will declared winner by default. The gesture detected in this case is a drag.
```
"will declared" => "will be declared"

justinmc commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -1870,6 +1858,23 @@ class TextSelectionGestureDetectorBuilder {
+  }
+
+  /// Returns true if shift left or right is contained in the given set.
+  bool _containsShift(Set<LogicalKeyboardKey> keysPressed) {
```
I think this can be static.

justinmc reviewed (approved):
LGTM with nits 👍

chunhtai commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -2011,21 +2019,18 @@ class TextSelectionGestureDetectorBuilder {
     // Handle shift + click selection if needed.
-    final bool isShiftPressedValid = _isShiftPressed && renderEditable.selection?.baseOffset != null;
+    final bool isShiftPressed = _containsShift(details.keysPressedOnDown);
+    final bool isShiftPressedValid = isShiftPressed && renderEditable.selection?.baseOffset != null;
```
It is a bit confusing for the name `isShiftPressedValid` can you rename it or add some comment on why the shiftpressed is only valid when there is a existing selection in renderEditable?

Renzo-Olivares replied:
Added a comment.

chunhtai reviewed (commented):
I think overall looks good, I don't have structural-wise cooments, all the comments are just minor questions in smaller scopes.

Renzo-Olivares commented on packages/flutter/lib/src/widgets/tap_and_drag_gestures.dart:
```diff
@@ -0,0 +1,1296 @@
+//   state as long as it continues to track a pointer.
+// * If a [PointerMoveEvent] is tracked that has moved a sufficient global distance
+//   from the initial [PointerDownEvent] and it came before a [PointerUpEvent], then
+//   this recognizer moves from the [possible] state to [accepted].
```
In the case described, when the sufficient global distance is met before a [PointerUpEvent] is received then the recognizer will declare self-victory.

Renzo-Olivares replied:
The cases when this recognizer declares victory are talked about in the `TapAndDragGestureRecognizer` docs.

Renzo-Olivares commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -2666,100 +2763,60 @@ class TextSelectionGestureDetector extends StatefulWidget {
+      case TargetPlatform.linux:
+      case TargetPlatform.macOS:
+      case TargetPlatform.windows:
+        return 2;
```
I plan on changing this with the triple tap to select PR. Some platforms seem to have varying behaviors in this regard.

Renzo-Olivares commented on packages/flutter/lib/src/widgets/tap_and_drag_gestures.dart:
```diff
@@ -0,0 +1,1296 @@
+        _giveUpPointer(event.pointer);
+      }
+
+      final bool isSlopPastTolerance = slopTolerance != null && _getGlobalDistance(event, _initialPosition) > slopTolerance!;
```
I think that would create conflicts. What if a `PointerMoveEvent` is received that is past the slop tolerance, but the next one received is not, then the `_pastSlopTolerance` would be reset. The expected behavior is that the `_pastTapTolerance` stay the same through the sequence of pointer events, until a `PointerUpEvent` is received or the recognizer loses or is disposed somehow.

chunhtai replied:
how about `_pastTapTolerance = _pastTapTolerance || ..`

Renzo-Olivares commented on packages/flutter/lib/src/widgets/tap_and_drag_gestures.dart:
```diff
@@ -0,0 +1,1296 @@
+class TapDragDownDetails {
+  /// Creates details for a [GestureTapDragDownCallback].
+  ///
+  /// The [globalPosition] argument must be provided and must not be null.
```
Sorry this was a mistake. `globalPosition` should be required on all details classes.

Commit:
Address some reviewer comments

Commit:
make deadline and slopTolerance private

Commit:
updates

Commit:
updates

Commit:
Address review comments

Renzo-Olivares commented on packages/flutter/lib/src/widgets/tap_and_drag_gestures.dart:
```diff
@@ -0,0 +1,1296 @@
+
+  bool _wonArena = false;
+  OffsetPair? _originPosition;
+  int? _previousButtons;
```
The _`TapStatusTrackerMixin` does not know if the taps are coming from the primary pointer. That knowledge is owned by `TapAndDragGestureRecognizer`. Say if someone wanted to have a recognizer that doesn't track a primary pointer, the `_TapStatusTrackerMixin` would still work to track all taps and not just the one from the primary pointer.

Renzo-Olivares commented on packages/flutter/lib/src/widgets/tap_and_drag_gestures.dart:
```diff
@@ -0,0 +1,1296 @@
+  @override
+  bool isPointerAllowed(PointerEvent event) {
+    if (_initialButtons == null) {
+      switch (event.buttons) {
```
got rid of `_initialButtons`.

Commit:
remove _initialButtons

Commit:
fix docs

Renzo-Olivares commented on packages/flutter/lib/src/widgets/tap_and_drag_gestures.dart:
```diff
@@ -0,0 +1,1296 @@
+    super.dispose();
+  }
+
+  void _trackTrap(PointerDownEvent event) {
```
Good catch! Not sure how I missed this one. Thanks!

Commit:
trackTrap -> trackTap

Commit:
fix analyzer

Renzo-Olivares commented on packages/flutter/lib/src/widgets/tap_and_drag_gestures.dart:
```diff
@@ -0,0 +1,1296 @@
+      }
+    } else if (event is PointerUpEvent) {
+      _up = event;
+      if (_wonArena && _down != null) {
```
This can happen if we haven't sent a tap down (a tap down is sent when the recognizer wins the arena or the `kPressTimeout` elapses) before arrive at a `PointerUpEvent` in `handleEvent`. This can happen when a recognizer is competing with this one because this recognizer will only declare self-victory if:

1. It is the last recognizer in the arena
2. The `kPressTimeout` has elapsed and the consecutive tap count is greater than 1.
3. The pointer has moved a sufficient global distance to be considered a drag (moved past the panSlop).

We only want to start the timer on tap up when we are sure a tap down has been sent.

Renzo-Olivares commented on packages/flutter/lib/src/widgets/tap_and_drag_gestures.dart:
```diff
@@ -0,0 +1,1296 @@
+      _tapTrackerReset();
+    }
+    _up = null;
+    _wonArena = false;
```
It is possible that `acceptGesture` is called without a `PointerUpEvent` being received. This can happen if the arena is sweeped before a `PointerUpEvent` is routed, and this recognizer is the only one in the arena.

Renzo-Olivares replied:
I added a test to verify this.

Renzo-Olivares replied:
I am for changing the name to `_isPointerDown` or `_sentTapDown` but am not sure about the `_isPointerDown = !(_up != null && _down != null);`.

chunhtai replied:
> _isPointerDown = !(_up != null && _down != null);

What I meant is the name also matches the logic in acceptGesture

`_isPointerDown = !(_up != null && _down != null);` is logically equivalent to

```dart
_isPointerDown = true;
if (_up != null && _down != null) {
 _isPointerDown= false;
}
```
assuming it is impossible that _up == null and _down ==null

Commit:
Add test to verify that tap up is called when recognizer accepts before handleEvent is called

Renzo-Olivares commented on packages/flutter/lib/src/widgets/tap_and_drag_gestures.dart:
```diff
@@ -0,0 +1,1296 @@
+  final Set<LogicalKeyboardKey> keysPressedOnDown;
+
+  @override
+  String toString() => '${objectRuntimeType(this, 'TapDragDownDetails')}($globalPosition)';
```
Used diagnosticable mixin instead.

Renzo-Olivares commented on packages/flutter/lib/src/widgets/tap_and_drag_gestures.dart:
```diff
@@ -0,0 +1,1296 @@
+  final Set<LogicalKeyboardKey> keysPressedOnDown;
+
+  @override
+  String toString() => '${objectRuntimeType(this, 'TapDragUpDetails')}($globalPosition)';
```
Used `Diagnosticable`.

Commit:
implement Diagnosticable for Details objects;

Commit:
sentTapDown == wonArenaForPrimaryPointer, so the implementation now only uses sentTapDown

Commit:
Count user tap up immediately and do not wait to win the arena

Commit:
Do not need to call super from TapAndDragGestureRecognizer.acceptGesture anymore because mixin implementation is gone

Commit:
Do not start selection drag on Android, iOS, and Fuchshsia touch devices if renderEditable does not have focus, this fixes many scubas

chunhtai reviewed (approved):
LGTM, just some minor comment

Renzo-Olivares commented on packages/flutter/lib/src/widgets/tap_and_drag_gestures.dart:
```diff
@@ -0,0 +1,1312 @@
+  /// See also:
+  ///
+  ///  * [DragGestureRecognizer.dragStartBehavior], which includes more details and an example.
+  DragStartBehavior dragStartBehavior;
```
Makes sense to me.

Commit:
Address reviewer comments

Commit:
fix test

Commit:
TapAndDragGestureRecognizer should wait for other recognizer to lose before winning the arena

chunhtai commented on packages/flutter/lib/src/widgets/tap_and_drag_gestures.dart:
```diff
@@ -0,0 +1,1285 @@
+/// screen) or a drag (e.g. if the pointer was not dragged far enough to
+/// be considered a drag.
+///
+/// It will not immediately declare victory for every tap or drag that it
```
```suggestion
/// This recognizer will not immediately declare victory for every tap or drag that it
```

chunhtai reviewed (approved):
still LGTM

Commit:
Address review comments

Commit:
Dont check for drag if the start was already found

Commit:
Only check for a drag if it has not already been found"

Commit:
fix from rebase
